src/skins/vector/views/organisms/UserSettings.js

```diff
@@ -15,15 +15,144 @@ limitations under the License.
+var version = require('../../../../../package.json').version;
 
 var Modal = require('matrix-react-sdk/lib/Modal');
+var UserSettingsStore = require('matrix-react-sdk/lib/UserSettingsStore');
 
 module.exports = React.createClass({
     displayName: 'UserSettings',
-    mixins: [UserSettingsController],
+
+    Phases: {
+        Loading: "loading",
+        Saving: "saving",
+        Display: "display",
+    },
+
+    getInitialState: function() {
+        return {
+            avatarUrl: null,
+            displayName: null,
+            threePids: [],
+            clientVersion: version,
+            phase: this.Phases.Loading,
+        };
+    },
+
+    componentWillMount: function() {
+        var self = this;
+
+        var profilePromise = UserSettingsStore.loadProfileInfo();
+        var threepidPromise = UserSettingsStore.loadThreePids();
+
+        q.all([profilePromise, threepidPromise]).then(
+            function(resps) {
+                self.setState({
+                    avatarUrl: resps[0].avatar_url,
+                    displayName: resps[0].displayname,
+                    threepids: resps[1].threepids,
+                    phase: self.Phases.Display,
+                });
+
+                // keep a copy of the original state in order to track changes
+                self.setState({
+                    originalState: self.state
+                });
+            },
+            function(error) {
+                var ErrorDialog = sdk.getComponent("organisms.ErrorDialog");
+                Modal.createDialog(ErrorDialog, {
+                    title: "Can't load user settings",
+                    description: error.toString()
+                });
+            }
+        );
+    },
+
+    componentDidMount: function() {
+        this.dispatcherRef = dis.register(this.onAction);
+    },
+
+    componentWillUnmount: function() {
+        dis.unregister(this.dispatcherRef);
+    },
+
+    onSaveClicked: function(ev) {
+        var self = this;
+        var savePromises = [];
+
+        // XXX: this is managed in ChangeAvatar.js, although could be moved out here in order
+        // to allow for the change to be staged alongside the rest of the form.
+        //
+        // if (this.state.originalState.avatarUrl !== this.state.avatarUrl) {
+        //     savePromises.push( UserSettingsStore.saveAvatarUrl(this.state.avatarUrl) );
+        // }
+
+        if (this.state.originalState.displayName !== this.state.displayName) {
+            savePromises.push( UserSettingsStore.saveDisplayName(this.state.displayName) );
+        }
+
+        if (this.state.originalState.threepids.length !== this.state.threepids.length ||
+            this.state.originalState.threepids.every(function(element, index) {
+                    return element === this.state.threepids[index];
+            }))
+        {
+            savePromises.push( UserSettingsStore.saveThreePids(this.state.threepids) );
```

Comment:
Why should this component be aware that there are 3 separate API calls which it needs to put into an array and then `.all()` on? If we ever update the HTTP API, we'd need to change this component *in addition* to the `UserSettingsStore`.

The act of lazily doing API calls iff there are changes is not necessarily a UI thing. Why can't this logic go into the store? This would simplify this component significantly if the exposed API was something like:
```javascript
var stuffToShow = UserSettingsStore.getUserSettings(); // returns object with all the things you may want to display; can optionally add in an arg to say "no really, I actually care about the API calls you're making, I only want 3PIDs".
this.setState({
  userSettings: stuffToShow
});
```

Which is then saved via:
```javascript
UserSettingsStore.saveSettings(this.state.userSettings); // this does the checks you're doing here
```